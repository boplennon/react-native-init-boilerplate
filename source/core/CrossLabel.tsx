import React from 'react';
import { View, TextProps, StyleSheet, Linking } from 'react-native';
import { Text } from 'react-native-paper';
import Theme from '../styles';
import _ from 'lodash';

export interface ICrossLabelProps extends TextProps {
  /**
   * An url to open when user clicks the label. Styles the label as a clickable "link"
   */
  onPressUrlTarget?: string;
}

/**
 * A custom text compoent that displays as a link if {@param onPressUrlTarget} is supplied
 *
 * @param param0 {@link TextProps}
 */
export class CrossLabel extends React.Component<ICrossLabelProps> {
  constructor(props: ICrossLabelProps) {
    super(props);
  }

  render() {
    const url = this.props.onPressUrlTarget;
    let style = this.props.style;

    let urlProp: TextProps = {};

    if (!_.isNil(url)) {
      urlProp = {
        onPress: () => Linking.openURL(url.toString()),
      };
      style = localStyles.link;
    }

    return (
      <View style={Theme.container}>
        <Text {...this.props} {...urlProp} style={style}>
          {this.props.children}
        </Text>
      </View>
    );
  }
}

const localStyles = StyleSheet.create({
  link: {
    color: 'blue',
    fontWeight: 'bold',
  },
});

export default CrossLabel;
