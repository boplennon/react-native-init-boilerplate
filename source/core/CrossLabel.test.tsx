// core/components/BusyIndicator.test
/// <reference types="jest"/>
import React from 'react';
import { shallow } from 'enzyme';
import CrossLabel from './CrossLabel';
import _ from 'lodash';

jest.unmock('react-native');

function setup(text:string, onPress?: () => void) {

  const enzymeWrapper = shallow(<CrossLabel onPress={!_.isNil(onPress)?onPress:jest.fn()}>{text}</CrossLabel>);

  return {
    enzymeWrapper,
  };
}

describe('components', () => {
  /**
   * Test component rendering. Properties of children might be tested by importing their type:
   *
   *    const busyProps = enzymeWrapper.find(BusyIndicator).props();
   *    expect(busyProps.isBusy).toBe(false);
   */
  describe('CrossLabel', () => {
    it('Without onPress should render', () => {
      const { enzymeWrapper } = setup('Hello');
      expect(enzymeWrapper).toMatchSnapshot();
    });

    it('With onPress should have link style', () => {
        const { enzymeWrapper } = setup('Click me', () =>{});
        expect(enzymeWrapper).toMatchSnapshot();
        // const text = enzymeWrapper.find(CrossLabel).at(0);
        // // @ts-ignore
        // expect(text.props().style).toEqual(styles.link);
      });
  });
});
