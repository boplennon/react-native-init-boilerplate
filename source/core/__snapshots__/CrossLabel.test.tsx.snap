// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components CrossLabel With onPress should have link style 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CrossLabel
    onPress={[Function]}
  >
    Click me
  </CrossLabel>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <UNDEFINED
        onPress={[Function]}
      >
        Click me
      </UNDEFINED>,
      "style": Object {
        "flex": 1,
        "justifyContent": "space-between",
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": "Click me",
        "onPress": [Function],
        "style": undefined,
      },
      "ref": null,
      "rendered": "Click me",
      "type": undefined,
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <UNDEFINED
          onPress={[Function]}
        >
          Click me
        </UNDEFINED>,
        "style": Object {
          "flex": 1,
          "justifyContent": "space-between",
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "Click me",
          "onPress": [Function],
          "style": undefined,
        },
        "ref": null,
        "rendered": "Click me",
        "type": undefined,
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`components CrossLabel Without onPress should render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CrossLabel
    onPress={[MockFunction]}
  >
    Hello
  </CrossLabel>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <UNDEFINED
        onPress={[MockFunction]}
      >
        Hello
      </UNDEFINED>,
      "style": Object {
        "flex": 1,
        "justifyContent": "space-between",
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": "Hello",
        "onPress": [MockFunction],
        "style": undefined,
      },
      "ref": null,
      "rendered": "Hello",
      "type": undefined,
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <UNDEFINED
          onPress={[MockFunction]}
        >
          Hello
        </UNDEFINED>,
        "style": Object {
          "flex": 1,
          "justifyContent": "space-between",
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "Hello",
          "onPress": [MockFunction],
          "style": undefined,
        },
        "ref": null,
        "rendered": "Hello",
        "type": undefined,
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
